### PR TITLE
fix(mariadb): json array filters

### PR DESCRIPTION
**Pull request by Weakky**

### CODE REVIEW

File: src/ast/function/json_extract_array.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+pub(crate) fn json_extract_first_array_elem<'a, E>(expr: E) -> Function<'a>
+where
+    E: Into<Expression<'a>>,
+{
+    let fun = JsonExtractFirstArrayElem {
+        expr: Box::new(expr.into()),
+    };
+
+    fun.into()
+}
```
Comment (Weakky) [inline src/ast/function/json_extract_array.rs:39]: I created two new _private_ functions to enable using the existing JSON equality code paths. See the comment below to see how they are used 👇

File: src/ast/expression.rs [hunk 2]
```diff
@@ -466,31 +466,39 @@ impl<'a> Comparable<'a> for Expression<'a> {
...
 
     #[cfg(all(feature = "json", any(feature = "postgresql", feature = "mysql")))]
     fn json_array_not_ends_into<T>(self, item: T) -> Compare<'a>
     where
         T: Into<Expression<'a>>,
     {
-        Compare::JsonCompare(JsonCompare::ArrayNotEndsInto(Box::new(self), Box::new(item.into())))
+        let array_ends_into: Expression = json_extract_last_array_elem(self).into();
+
+        Compare::NotEquals(Box::new(array_ends_into), Box::new(item.into()))
```
Comment (Weakky) [inline src/ast/expression.rs:501]: The private functions are used here 👆.

Instead of having special `JsonCompare`s, we rely on the existing `Compare::Equals` & `Compare::NotEquals`.

The private functions help us _wrap_ the left side of the equal into whatever's needed to get the first/last item of an array.

File: src/ast/compare.rs [hunk 3]
```diff
@@ -59,10 +59,6 @@ pub enum Compare<'a> {
 pub enum JsonCompare<'a> {
     ArrayContains(Box<Expression<'a>>, Box<Expression<'a>>),
     ArrayNotContains(Box<Expression<'a>>, Box<Expression<'a>>),
-    ArrayBeginsWith(Box<Expression<'a>>, Box<Expression<'a>>),
-    ArrayNotBeginsWith(Box<Expression<'a>>, Box<Expression<'a>>),
-    ArrayEndsInto(Box<Expression<'a>>, Box<Expression<'a>>),
-    ArrayNotEndsInto(Box<Expression<'a>>, Box<Expression<'a>>),
```
Comment (Weakky) [inline src/ast/compare.rs:65]: As explained above, this means we no longer need these `JsonCompare` variants as we now rely on `Compare::Equals` & `Compare::NotEquals`.

File: src/visitor.rs [hunk 4]
```diff
@@ -872,10 +872,6 @@ pub trait Visitor<'a> {
             Compare::JsonCompare(json_compare) => match json_compare {
                 JsonCompare::ArrayContains(left, right) => self.visit_json_array_contains(*left, *right, false),
                 JsonCompare::ArrayNotContains(left, right) => self.visit_json_array_contains(*left, *right, true),
-                JsonCompare::ArrayBeginsWith(left, right) => self.visit_json_array_begins_with(*left, *right, false),
-                JsonCompare::ArrayNotBeginsWith(left, right) => self.visit_json_array_begins_with(*left, *right, true),
-                JsonCompare::ArrayEndsInto(left, right) => self.visit_json_array_ends_into(*left, *right, false),
-                JsonCompare::ArrayNotEndsInto(left, right) => self.visit_json_array_ends_into(*left, *right, true),
```
Comment (Weakky) [inline src/visitor.rs:878]: As explained above though, we could as well keep the `JsonCompare` variants and create the `Compare::Equals|NotEquals` variants in here if that's preferable for visibility.

File: src/tests/test_api.rs [hunk 5]
```diff
@@ -25,6 +27,7 @@ pub trait TestApi {
     async fn create_index(&mut self, table: &str, columns: &str) -> crate::Result<String>;
 
     fn system(&self) -> &'static str;
+    fn connector_tag(&self) -> Tags;
```
Comment (Weakky) [inline src/tests/test_api.rs:30]: Last thing: unfortunately, `JSON_CONTAINS` doesn't behave the same on MariaDB as it does on MySQL. This was causing issues on test assertions. I, therefore, added a new method to enable getting the current tag of the connector. See the comment below how it's being used 👇🏻

File: src/tests/query.rs [hunk 6]
```diff
@@ -2111,36 +2113,52 @@ async fn json_array_contains_fun(api: &mut dyn TestApi) -> crate::Result<()> {
...
+            value_into_json(&res.next().unwrap()["obj"])
+        );
+        assert_eq!(None, res.next());
+    } else {
+        assert_eq!(
+            Some(serde_json::json!({ "a": { "b": [[1, 2], [3, 4]] } })),
+            value_into_json(&res.next().unwrap()["obj"])
+        );
+        assert_eq!(None, res.next());
+    }
```
Comment (Weakky) [inline src/tests/query.rs:2161]: This is where we branch the assertions in the case of MariaDB

**Review by pimeys — APPROVED** [review]
lgtm

Hopefully no more weird JSON edge cases after this :)
